### PR TITLE
Support init accessor in CSharpSyntaxFacts

### PR DESCRIPTION
**Pull request by Youssef1313**
Prefer squash/merge.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs [hunk 1]
```diff
@@ -2006,11 +2008,11 @@ public override DeclarationKind GetDeclarationKind(SyntaxNode declaration)
                         return DeclarationKind.Attribute;
                     }
                     break;
-
+                // Add InitAccessor once https://github.com/dotnet/roslyn/issues/48136 is resolved
                 case SyntaxKind.GetAccessorDeclaration:
                     return DeclarationKind.GetAccessor;
                 case SyntaxKind.SetAccessorDeclaration:
-                    return DeclarationKind.SetAccessor;
+                    return DeclarationKind.SetAccessor;:
```
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:2015]:
```suggestion
                    return DeclarationKind.SetAccessor;
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs [hunk 2]
```diff
@@ -2006,7 +2008,7 @@ public override DeclarationKind GetDeclarationKind(SyntaxNode declaration)
                         return DeclarationKind.Attribute;
                     }
                     break;
-
+                // Add InitAccessor once https://github.com/dotnet/roslyn/issues/48136 is resolved
```
Comment (davidwengier) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:2011]: I don't think this comment is adding any value.

Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:2011]:
```suggestion

```

Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs]: Removed it.

**Review by davidwengier — APPROVED** [review]
This change makes sense, but I don't see too much value in aggressively updating all of these APIs without known consuming use cases, or tests (or both!).

**Comment by Youssef1313** [comment]
@davidwengier I looked at the usages of one of these changes, and it ends up being used for the public API SyntaxGenerator.WithAccessibility.

It may not be needed internally for roslyn *currently*. But consumers of this API should get the desired behavior.

I haven't confirmed whether the other change ends up being used in a public API or not.

**Comment by CyrusNajmabadi** [comment]
> and it ends up being used for the public API SyntaxGenerator.WithAccessibility.

THis is a testable scenario :)  Plz add test.

**Comment by davidwengier** [comment]
> It may not be needed internally for roslyn currently. But consumers of this API should get the desired behavior.

I don't disagree, but there could also be issues if things light up in areas that they were never intended to, or otherwise don't support. Hopefully, of course, the existing test coverage validates that :)

**Comment by Youssef1313** [comment]
@davidwengier Got it. Let me check what tests will fail in #48158 and update these tests for init accessor in this PR.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs [hunk 3]
```diff
@@ -1315,6 +1315,7 @@ private static bool IsMemberDeclaration(SyntaxNode node)
                 case SyntaxKind.PropertyDeclaration:
                 case SyntaxKind.GetAccessorDeclaration:
                 case SyntaxKind.SetAccessorDeclaration:
+                case SyntaxKind.InitAccessorDeclaration:
```
Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:1318]: @davidwengier I removed both GetAccessorDeclaration and InitAccessorDeclaration in a draft. There are no test failures. So the current tests, unfortunately, doesn't cover those.

The usages that depend on this method are in:

- CodeLens\CodeLensReferencesService.cs (there is a call to `syntaxFactsService.IsDeclaration(node)`)
- CodeRefactorings\SyncNamespace\AbstractChangeNamespaceService.cs (there is a call to `.DescendantNodes(n => !syntaxFacts.IsDeclaration(n))`)

Unlike the other change, I'm not able to guess the effect of this change.
I'll try to change the method to only `return false` in the draft in a hope to see test failures (i.e. other switch arms are tested).

Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:1318]: I don't think it's a good idea to update this one without knowing the change it introduce.

This **whole** method isn't covered in the tests. See how https://github.com/dotnet/roslyn/pull/48158 is green.

Comment (Youssef1313) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SyntaxFacts/CSharpSyntaxFacts.cs:1318]: Reverting, and will file an issue to track test coverage for this method and adding the init accessor if necessary.

```suggestion
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by davidwengier — APPROVED** [review]

**Comment by Youssef1313** [comment]
@CyrusNajmabadi @davidwengier Is this ready to merge?

**Comment by davidwengier** [comment]
@sharwell

**Review by sharwell — APPROVED** [review]

**Pull request merged**